pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -115,6 +115,10 @@ func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reserv
 	if err != nil {
 		return nil, err
 	}
+	err = ValidateStaticPolicyOptions(opts, topology, affinity)
```

Comment:
I do need feedback here. 
Based on my understanding, it ensure that policy option code is located at just one place otherwise we have to generate topology in container manager.
I understand there is early panic pattern in kubelet. This code appears to be in initial bootstrap path so it should ideally result in early failure

Reply:
Right, I'll spend a bit more time here and provided a more targeted suggestion.

Reply:
I had a closer look. There are intertwined deps, and I can't see yet a clear way to de-entangle here.
1. checking the topology belongs in the cpumanager indeed, fair enough about that
2. checking the topology manager configuration, especially inferring the behaviour from the name, gives howver a bad taste.

Moving the validation a level above would require changes on how we parse/create the StaticPolicyOptions (possibly renaming them) *and* inferring the cputopology from the machineinfo, which is probably too much - and wasted work for the `none` policy. So atm this direction doesn't seem much better.

In general, we have now the need to validate the value of the option vs the configuration of _other_ managers - namely the topology manager - while previously we only had the need to validate the options against each other, fully inside the cpumanager (well, the options were compatible with each other anyway so no big deal, but still).

Overall, I think the best possible improvement for the alpha level would be to add a method to the `topologymanager.Policy` interface to make the validation against the behaviour - not the name.

Last but not least, after this check I don't think this change is alpha-blocking. Let's try to make it happen, but if we can't reach a clear improvement I think we can postpone to beta. If we postpone, in the beta stage we should really give a good look at things here.

Reply:
I have no major objections with the approach taken here but I was thinking of an alternative approach.

We need this validation specifically when `AlignBySocket` option is enabled then why don't we make the validation more targeted towards CPU Manager policy options (a level down) rather than in policy_static.go.

I suggest that we change `ValidateStaticPolicyOptions` to `ValidateAlignBySocketOption` and do this validation akin to  `CheckPolicyOptionAvailable` and make changes like below:
```
diff --git a/pkg/kubelet/cm/cpumanager/policy_options.go b/pkg/kubelet/cm/cpumanager/policy_options.go
index 775470bae99..8958e732398 100644
--- a/pkg/kubelet/cm/cpumanager/policy_options.go
+++ b/pkg/kubelet/cm/cpumanager/policy_options.go
@@ -78,7 +78,7 @@ type StaticPolicyOptions struct {
        AlignBySocket bool
 }
 
-func NewStaticPolicyOptions(policyOptions map[string]string) (StaticPolicyOptions, error) {
+func NewStaticPolicyOptions(policyOptions map[string]string, topology *topology.CPUTopology, topologyManagerPolicy string) (StaticPolicyOptions, error) {
        opts := StaticPolicyOptions{}
        for name, value := range policyOptions {
                if err := CheckPolicyOptionAvailable(name); err != nil {
@@ -99,6 +99,9 @@ func NewStaticPolicyOptions(policyOptions map[string]string) (StaticPolicyOption
                        }
                        opts.DistributeCPUsAcrossNUMA = optValue
                case AlignBySocketOption:
+                       if err := ValidateAlignBySocketOption(opts, topology, topologyManagerPolicy); err != nil {
+                               return opts, err
+                       }
                        optValue, err := strconv.ParseBool(value)
                        if err != nil {
                                return opts, fmt.Errorf("bad value for option %q: %w", name, err)
@@ -113,16 +116,14 @@ func NewStaticPolicyOptions(policyOptions map[string]string) (StaticPolicyOption
        return opts, nil
 }
 
-func ValidateStaticPolicyOptions(opts StaticPolicyOptions, topology *topology.CPUTopology, topologyManager topologymanager.Store) error {
-       if opts.AlignBySocket == true {
-               //1. Not compatible with topology manager single-numa-node policy option.
-               if topologyManager.GetPolicy().Name() == topologymanager.PolicySingleNumaNode {
-                       return fmt.Errorf("Topolgy manager Single numa policy is incompatible with CPUManager Align  by socket policy option")
-               }
-               //2. Not compatible with topology when number of sockets are more than number of NUMA nodes.
-               if topology.NumSockets > topology.NumNUMANodes {
-                       return fmt.Errorf("Align by socket is not compatible with hardware where number of sockets are more than number of NUMA")
-               }
+func ValidateAlignBySocketOption(opts StaticPolicyOptions, topology *topology.CPUTopology, topologyManagerPolicy string) error {
+       //1. Not compatible with topology manager single-numa-node policy option.
+       if topologyManagerPolicy == topologymanager.PolicySingleNumaNode {
+               return fmt.Errorf("Topolgy manager Single numa policy is incompatible with CPUManager Align  by socket policy option")
+       }
+       //2. Not compatible with topology when number of sockets are more than number of NUMA nodes.
+       if topology.NumSockets > topology.NumNUMANodes {
+               return fmt.Errorf("Align by socket is not compatible with hardware where number of sockets are more than number of NUMA")
        }
        return nil
 }
diff --git a/pkg/kubelet/cm/cpumanager/policy_static.go b/pkg/kubelet/cm/cpumanager/policy_static.go
index 4ed9fdc0132..d0b9ec73068 100644
--- a/pkg/kubelet/cm/cpumanager/policy_static.go
+++ b/pkg/kubelet/cm/cpumanager/policy_static.go
@@ -111,11 +111,7 @@ var _ Policy = &staticPolicy{}
 // assignments for exclusively pinned guaranteed containers after the main
 // container process starts.
 func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store, cpuPolicyOptions map[string]string) (Policy, error) {
-       opts, err := NewStaticPolicyOptions(cpuPolicyOptions)
-       if err != nil {
-               return nil, err
-       }
-       err = ValidateStaticPolicyOptions(opts, topology, affinity)
+       opts, err := NewStaticPolicyOptions(cpuPolicyOptions, topology, affinity.GetPolicy().Name())
        if err != nil {
                return nil, err
        }
```
The only possible bone of contention might be that we have to pass topology and topologymanager policy to NewStaticPolicyOptions but we pass the topology information to CPU Manager with static policy anyway, I don't see a problem in passing it further to StaticPolicyOptions for validation. Unless passing node topology and topologyManagerPolicy is seen wrong (which I am unable to see at this stage and I am open to be corrected), I think this could be a good approach. What we gain is more coherent and readable code.

@fromanirh @arpitsardhana  WDYT?

Reply:
I see your point here @swatisehgal . You approach has merit.
What bothers me is deepening the coupling between cpumanager and topology manager. Passing these parameters one level below increases the coupling a bit. The approach I was evaluating - moving the validation one level *above* - was tring to reduce, possibly eliminate the additional coupling.

However, I don't have a clear solution to propose, so I'm neutral here. We should totally spend some time before Beta, though.

Reply:
I'm brainstorming a bit, again I want to stress I don't have a well formed opinion here so I'm not blocking for alpha here.
Indeed the TopologyManager collects hints and should make the final call here, being the orchestrator. Doing the check inside a provider -point in case cpumanager- feels backward.
So I think a possible solution lies in the `HintProvider` interface. We can add a method to signal a specific configuration of a specific provider cannot meet an alignment criteria (point in case: single numa node). The orchestrator -topology manager - can do a quick pass over all the registered hintproviders and check if they all meet its own configuration requirement (= they are all compatible with the configured policy, point in case single-numa-node), Then it's up to the orchestrator (topology manager) how to handle this case. I think we should bail out and fail.

Reply:
Demo of what I mean: partial patch, untested, for illustration purposes only:
```
diff --git a/pkg/kubelet/cm/container_manager_linux.go b/pkg/kubelet/cm/container_manager_linux.go
index ecf4d89c1eb..a548fb6d157 100644
--- a/pkg/kubelet/cm/container_manager_linux.go
+++ b/pkg/kubelet/cm/container_manager_linux.go
@@ -345,6 +345,9 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
 		cm.topologyManager.AddHintProvider(cm.memoryManager)
 	}
 
+	if err := cm.topologyManager.ValidateHintProviders(); err != nil {
+		return nil, err
+	}
 	return cm, nil
 }
 
diff --git a/pkg/kubelet/cm/cpumanager/cpu_manager.go b/pkg/kubelet/cm/cpumanager/cpu_manager.go
index dde49b6ec8c..f843685e539 100644
--- a/pkg/kubelet/cm/cpumanager/cpu_manager.go
+++ b/pkg/kubelet/cm/cpumanager/cpu_manager.go
@@ -334,6 +334,10 @@ func (m *manager) GetPodTopologyHints(pod *v1.Pod) map[string][]topologymanager.
 	return m.policy.GetPodTopologyHints(m.state, pod)
 }
 
+func (m *manager) GetAlignmentOptions() topologymanager.AlignmentOptions {
+	return m.policy.GetAlignmentOptions()
+}
+
 func (m *manager) GetAllocatableCPUs() cpuset.CPUSet {
 	return m.allocatableCPUs.Clone()
 }
diff --git a/pkg/kubelet/cm/cpumanager/policy_none.go b/pkg/kubelet/cm/cpumanager/policy_none.go
index 1e35f6a094e..f1f66c9d3aa 100644
--- a/pkg/kubelet/cm/cpumanager/policy_none.go
+++ b/pkg/kubelet/cm/cpumanager/policy_none.go
@@ -19,7 +19,7 @@ package cpumanager
 import (
 	"fmt"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
 	"k8s.io/klog/v2"
 	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/state"
 	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
@@ -66,6 +66,10 @@ func (p *nonePolicy) GetPodTopologyHints(s state.State, pod *v1.Pod) map[string]
 	return nil
 }
 
+func (p *nonePolicy) GetAlignmentOptions() topologymanager.AlignmentOptions {
+	return topologymanager.MachineAlignment
+}
+
 // Assignable CPUs are the ones that can be exclusively allocated to pods that meet the exclusivity requirement
 // (ie guaranteed QoS class and integral CPU request).
 // Assignability of CPUs as a concept is only applicable in case of static policy i.e. scenarios where workloads
diff --git a/pkg/kubelet/cm/cpumanager/policy_static.go b/pkg/kubelet/cm/cpumanager/policy_static.go
index 4ed9fdc0132..871a935bdde 100644
--- a/pkg/kubelet/cm/cpumanager/policy_static.go
+++ b/pkg/kubelet/cm/cpumanager/policy_static.go
@@ -530,6 +530,15 @@ func (p *staticPolicy) GetPodTopologyHints(s state.State, pod *v1.Pod) map[strin
 	}
 }
 
+func (p *staticPolicy) GetAlignmentOptions() topologymanager.AlignmentOptions {
+	if p.options.AlignBySocket {
+		// we can't anymore align by NUMA node, and the HintProvider collector
+		// doesn't care yet about aligning by socket
+		return topologymanager.MachineAlignment
+	}
+	return topologymanager.NUMANodeAlignment
+}
+
 // generateCPUtopologyHints generates a set of TopologyHints given the set of
 // available CPUs and the number of CPUs being requested.
 //
diff --git a/pkg/kubelet/cm/topologymanager/policy.go b/pkg/kubelet/cm/topologymanager/policy.go
index 40255ed95cc..e3d01e39b20 100644
--- a/pkg/kubelet/cm/topologymanager/policy.go
+++ b/pkg/kubelet/cm/topologymanager/policy.go
@@ -28,6 +28,8 @@ type Policy interface {
 	// Returns a merged TopologyHint based on input from hint providers
 	// and a Pod Admit Handler Response based on hints and policy type
 	Merge(providersHints []map[string][]TopologyHint) (TopologyHint, bool)
+	// RequiredAlignment returns the alignment this policy enforces
+	Alignment() AlignmentOptions
 }
 
 // Merge a TopologyHints permutation to a single hint by performing a bitwise-AND
diff --git a/pkg/kubelet/cm/topologymanager/scope.go b/pkg/kubelet/cm/topologymanager/scope.go
index ed149df5bad..4ad89810491 100644
--- a/pkg/kubelet/cm/topologymanager/scope.go
+++ b/pkg/kubelet/cm/topologymanager/scope.go
@@ -17,9 +17,10 @@ limitations under the License.
 package topologymanager
 
 import (
+	"fmt"
 	"sync"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
 	"k8s.io/klog/v2"
 	"k8s.io/kubernetes/pkg/kubelet/cm/admission"
 	"k8s.io/kubernetes/pkg/kubelet/cm/containermap"
@@ -97,6 +98,16 @@ func (s *scope) AddHintProvider(h HintProvider) {
 	s.hintProviders = append(s.hintProviders, h)
 }
 
+func (s *scope) ValidateHintProviders() error {
+	alignment := s.policy.Alignment()
+	for _, provider := range s.hintProviders {
+		if (alignment & provider.GetAlignmentOptions()) == 0 {
+			return fmt.Errorf("registered hintprovider incompatible with policy alignment requirement")
+		}
+	}
+	return nil
+}
+
 // It would be better to implement this function in topologymanager instead of scope
 // but topologymanager do not track mapping anymore
 func (s *scope) AddContainer(pod *v1.Pod, container *v1.Container, containerID string) {
diff --git a/pkg/kubelet/cm/topologymanager/topology_manager.go b/pkg/kubelet/cm/topologymanager/topology_manager.go
index 8d11c7a5b88..1628ed165f1 100644
--- a/pkg/kubelet/cm/topologymanager/topology_manager.go
+++ b/pkg/kubelet/cm/topologymanager/topology_manager.go
@@ -20,7 +20,7 @@ import (
 	"fmt"
 
 	cadvisorapi "github.com/google/cadvisor/info/v1"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
 	"k8s.io/klog/v2"
 	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
 	"k8s.io/kubernetes/pkg/kubelet/lifecycle"
@@ -62,6 +62,9 @@ type Manager interface {
 	AddContainer(pod *v1.Pod, container *v1.Container, containerID string)
 	// RemoveContainer removes pod from Manager tracking
 	RemoveContainer(containerID string) error
+	// ValidateHintProviders verify that the capabilities of the registered
+	// HintProviders match the Manager configuration
+	ValidateHintProviders() error
 	// Store is the interface for storing pod topology hints
 	Store
 }
@@ -71,6 +74,13 @@ type manager struct {
 	scope Scope
 }
 
+type AlignmentOptions uint32
+
+const (
+	MachineAligment   AlignmentOptions = 0 // aka no sub-alignment
+	NUMANodeAlignment AlignmentOptions = 1 << iota
+)
+
 // HintProvider is an interface for components that want to collaborate to
 // achieve globally optimal concrete resource alignment with respect to
 // NUMA locality.
@@ -90,6 +100,10 @@ type HintProvider interface {
 	// all hints have been gathered and the aggregated Hint is available via a
 	// call to Store.GetAffinity().
 	Allocate(pod *v1.Pod, container *v1.Container) error
+	// GetAlignmentsOptions returns the capabilities of this hint provider
+	// in terms of alignment. Some hint providers, depending on their configuration
+	// lose the capability to align per NUMA zone.
+	GetAlignmentOptions() AlignmentOptions
 }
 
 //Store interface is to allow Hint Providers to retrieve pod affinity
@@ -193,6 +207,10 @@ func (m *manager) AddHintProvider(h HintProvider) {
 	m.scope.AddHintProvider(h)
 }
 
+func (m *manager) ValidateHintProviders() error {
+	return m.scope.ValidateHintProviders()
+}
+
 func (m *manager) AddContainer(pod *v1.Pod, container *v1.Container, containerID string) {
 	m.scope.AddContainer(pod, container, containerID)
 }
```

Reply:
Thanks a lot @swatisehgal and @fromanirh for extensive feedback.
I will experiment with both of above suggested methodologies and judge pros/cons.

There are three possibilities as we have listed here:
1. Top Level check: Doing it in hintprovider interface as suggested by Francesco
            a. Pros: More clean interface
            b. Cons: May be bit more complex, We already have coupling between topology manager & cpu manager with TopologyManager.Store() interface

2. Static Policy level check: As what i am currently doing
        a. Pros: One umbrella check for policy validation, can be used in future by other policy options
        b. Cons: Require defining new topology manager Store function thereby increasing coupling b/w topology & cpu manager, Validating all policy options check at one function may be cumbersome

3. Policy options check:  Doing it a level down from static policy in policy options  code path where we instantiate alignBySocket
              a. Pros: Targeted check to policy option, cleaner code
               b. Cons: Does not support check if multiple policy option together need to be evaluated. Also coupling increased between topology and cpu manager

I will get more clarity with implementation of above approach, so will update.

Reply:
I think part of the problem here is that this new flag probably belongs on the topologymanager as a "topology-manager-policy-option", not the CPUManager (for which we don't have any infrastructure yet). Let's leave it as is for the time being until we have time to evaluate if it makes sense moving this flag to the topologymanager in a future release (which may just have the opposite problem for the 2nd check going on which is CPUManager specific).

Reply:
I am leaving Validation as it is based on above discussion.
We can take close look during beta phase on how & where we should do Validation keeping above feedback in mind.

@swatisehgal I do see your point and experimented with suggested code. I wrote Validation function with thought that it can be used to add validation of existing and any future policy option. With individual policy option specific validation, we will anyway need a function for combined validation

Nevertheless to Kevin point, One of point to consider in beta is whether we want the policy option in CPU Manager or determine if Topology manager is better place if  unified alignment of device, memory & cpu manager is priority.